types/assoc.d.ts

```diff
@@ -1,7 +1,10 @@
 import { Placeholder, AssocPartialOne } from './util/tools';
 
-export function assoc<T, U>(__: Placeholder, val: T, obj: U): <K extends string>(prop: K) => Record<K, T> & Omit<U, K>;
-export function assoc<U, K extends string>(prop: K, __: Placeholder, obj: U): <T>(val: T) => Record<K, T> & Omit<U, K>;
-export function assoc<T, U, K extends string>(prop: K, val: T, obj: U): Record<K, T> & Omit<U, K>;
+export function assoc<T, U>(__: Placeholder, val: T, obj: U): <K extends string>(prop: K) => K extends keyof U ? T extends U[K] ? U : Record<K, T> & Omit<U, K> : U & Record<K, T>;
+export function assoc<U, K extends keyof U>(prop: K, __: Placeholder, obj: U): <T>(val: T) => T extends U[K] ? U : Record<K, T> & Omit<U, K>;
+export function assoc<U, K extends string>(prop: K, __: Placeholder, obj: U): <T>(val: T) => U & Record<K, T>;
+export function assoc<K extends keyof U, U>(prop: K, val: U[K], obj: U): U;
```

Comment:
Is this covered by the next definition? ie `export function assoc<T, K extends keyof U, U>(prop: K, val: T, obj: U): Record<K, T> & Omit<U, K>;`

Is there a reason to have both?

Reply:
I looked back at how I wrote this, and with some things I learned from doing my other MRs, I completely rewrote it. It has a lot more comments now so it'll be easier to review. I also was able to remove the `AssocPartialOne` util type, it's no longer needed